Mathlib/Algebra/Lie/LieTheorem.lean

```diff
@@ -0,0 +1,240 @@
+/-
+Copyright (c) 2024 Lucas Whitfield. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lucas Whitfield, Johan Commelin
+-/
+import Mathlib.Algebra.Lie.Weights.Basic
+import Mathlib.RingTheory.Finiteness.Nilpotent
+
+/-!
+# Lie's theorem for Solvable Lie algebras.
+
+Lie's theorem asserts that Lie modules of solvable Lie algebras over fields of characteristic 0
+have a common eigenvector for the action of all elements of the Lie algebra.
+This result is named `LieModule.exists_forall_lie_eq_smul_of_isSolvable`.
+-/
+
+namespace LieModule
+
+section
+
+/-
+The following variables generalize the setting where:
+- `R` is a principal ideal domain of characteristic zero,
+- `L` is a Lie algebra over `R`,
+- `V` is a Lie algebra module over `L`
+- `A` is a Lie ideal of `L`.
+Besides generalizing, it also make the proof of `lie_stable` syntactically smoother.
+-/
+variable {R L A V : Type*} [CommRing R]
+variable [IsPrincipalIdealRing R] [IsDomain R] [CharZero R]
+variable [LieRing L] [LieAlgebra R L]
+variable [LieRing A] [LieAlgebra R A]
+variable [LieRingModule L A] [LieRingModule A L]
+variable [AddCommGroup V] [Module R V] [Module.Free R V] [IsNoetherian R V]
+variable [LieRingModule L V] [LieModule R L V]
+variable [LieRingModule A V] [LieModule R A V]
+variable [IsLieTower L A V] [IsLieTower A L V]
+
+variable (χ : A → R)
+
+open Module (finrank)
+open LieModule
+
+local notation "π" => LieModule.toEnd R _ V
+
+private abbrev T (w : A) : Module.End R V := (π w) - χ w • 1
+
+lemma lie_stable (z : L) (v : V) (hv : v ∈ weightSpace V χ) : ⁅z, v⁆ ∈ weightSpace V χ := by
```

Comment:
How about
```suggestion
/-- An auxiliary lemma used only in the definition `LieModule.weightSpaceOfIsLieTower` below. -/
private lemma weightSpaceOfIsLieTower_aux (z : L) (v : V) (hv : v ∈ weightSpace V χ) :
    ⁅z, v⁆ ∈ weightSpace V χ := by
```
and then adding:
```lean
/-- The weight space of `V` with respect to `χ : A → R`, a priori a Lie submodule for `A`, is also a
Lie submodule for `L`. -/
def weightSpaceOfIsLieTower (χ : A → R) : LieSubmodule R L V :=
  { toSubmodule := weightSpace V χ
    lie_mem {z v} hv := weightSpaceOfIsLieTower_aux χ z v hv }
```

Reply:
I'm worried that there are very few applications for this definition. So I would prefer to add this once there is more than one use case.

Reply:
I think there's a good case for doing this.

For one thing, it makes it much easier to see the issue (`A`-weight space is `L`-submodule) at a glance. But more importantly it makes it possible to use join up the main result `lie_stable` with the `Liesubmodule` API, and indeed I note that this is exactly what is done where it is used.